File: platforms/android/foundation/foundation/src/main/java/io/nativeblocks/foundation/util/GradientUtil.kt
package io.nativeblocks.foundation.util

import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.LinearGradientShader
import androidx.compose.ui.graphics.Shader
import androidx.compose.ui.graphics.ShaderBrush
import androidx.core.graphics.toColorInt
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

internal const val GRADIENT_LINEAR = "linear"
internal const val GRADIENT_RADIAL = "radial"
internal const val GRADIENT_SWEEP = "sweep"

/**
 * Parses a comma-separated list of hexadecimal colors (e.g. "#004FF0, #00FFFFFF"),
 * dropping entries that fail to parse.
 */
internal fun parseColors(colors: String): List<Color> {
    return colors.split(",").mapNotNull { value ->
        runCatching { Color(value.trim().toColorInt()) }.getOrNull()
    }
}

/**
 * Builds a gradient [Brush] of the given type. Linear gradients honor [angleDegrees],
 * where 0 flows start-to-end and 90 flows top-to-bottom.
 */
internal fun gradientBrush(type: String, colors: List<Color>, angleDegrees: Float): Brush {
    return when (type) {
        GRADIENT_RADIAL -> Brush.radialGradient(colors)
        GRADIENT_SWEEP -> Brush.sweepGradient(colors)
        else -> AngularLinearGradient(colors, angleDegrees)
    }
}

/**
 * A linear gradient that flows at an arbitrary angle across the drawing area,
 * since [Brush.linearGradient] only supports fixed start/end offsets.
 */
internal data class AngularLinearGradient(
    private val colors: List<Color>,
    private val angleDegrees: Float,
) : ShaderBrush() {

    override fun createShader(size: Size): Shader {
        val angleRadians = Math.toRadians(angleDegrees.toDouble())
        val direction = Offset(cos(angleRadians).toFloat(), sin(angleRadians).toFloat())
        // half-length of the rect's projection onto the gradient axis
        val halfLength = (abs(direction.x) * size.width + abs(direction.y) * size.height) / 2f
        val center = Offset(size.width / 2f, size.height / 2f)
        return LinearGradientShader(
            from = center - Offset(direction.x * halfLength, direction.y * halfLength),
            to = center + Offset(direction.x * halfLength, direction.y * halfLength),
            colors = colors,
        )
    }
}

File: platforms/ios/foundation/Sources/NativeblocksFoundation/util/GradientUtil.swift
import Foundation
import SwiftUI

let GRADIENT_LINEAR = "linear"
let GRADIENT_RADIAL = "radial"
let GRADIENT_SWEEP = "sweep"

/// Parses a comma-separated list of hexadecimal colors (e.g. "#004FF0, #00FFFFFF"),
/// dropping entries that fail to parse.
func parseColors(_ colors: String) -> [Color] {
    return colors.split(separator: ",").compactMap { parseColor(String($0)) }
}

func parseColor(_ value: String) -> Color? {
    let hex = value.trimmingCharacters(in: .whitespacesAndNewlines).replacingOccurrences(of: "#", with: "")
    guard hex.count == 6 || hex.count == 8, hex.allSatisfy({ $0.isHexDigit }) else { return nil }

    var rgb: UInt64 = 0
    guard Scanner(string: hex).scanHexInt64(&rgb) else { return nil }

    if hex.count == 6 {
        return Color(
            red: CGFloat((rgb & 0xFF0000) >> 16) / 255.0,
            green: CGFloat((rgb & 0x00FF00) >> 8) / 255.0,
            blue: CGFloat(rgb & 0x0000FF) / 255.0
        )
    }
    return Color(
        red: CGFloat((rgb & 0x00FF_0000) >> 16) / 255.0,
        green: CGFloat((rgb & 0x0000_FF00) >> 8) / 255.0,
        blue: CGFloat(rgb & 0x0000_00FF) / 255.0,
        opacity: CGFloat((rgb & 0xFF00_0000) >> 24) / 255.0
    )
}

/// Builds a gradient style of the given type. Linear gradients honor `angleDegrees`,
/// where 0 flows start-to-end and 90 flows top-to-bottom.
func gradientStyle(_ type: String, _ colors: [Color], _ angleDegrees: Double) -> AnyShapeStyle {
    switch type {
    case GRADIENT_RADIAL:
        return AnyShapeStyle(EllipticalGradient(colors: colors))
    case GRADIENT_SWEEP:
        return AnyShapeStyle(AngularGradient(colors: colors, center: .center))
    default:
        let radians = angleDegrees * .pi / 180
        let dx = cos(radians) / 2
        let dy = sin(radians) / 2
        return AnyShapeStyle(
            LinearGradient(
                colors: colors,
                startPoint: UnitPoint(x: 0.5 - dx, y: 0.5 - dy),
                endPoint: UnitPoint(x: 0.5 + dx, y: 0.5 + dy)
            )
        )
    }
}
